metadata-ingestion/src/datahub/sdk/lineage_client.py

```diff
@@ -113,6 +126,286 @@ def normalize(s: str) -> str:
+        """
+        ...
+
+    @overload
+    def add_lineage(
+        self,
+        *,
+        upstream: DatasetUrnOrStr,
+        downstream: DatajobUrnOrStr,
+    ) -> None:
+        """
+        Add dataset to datajob lineage (dataset as input to job).
+
+        Args:
+            upstream: URN of the upstream dataset
+            downstream: URN of the downstream datajob
+        """
+        ...
+
+    @overload
+    def add_lineage(
+        self,
+        *,
+        upstream: DatajobUrnOrStr,
+        downstream: DatasetUrnOrStr,
+    ) -> None:
+        """
+        Add datajob to dataset lineage (dataset as output of job).
+
+        Args:
+            upstream: URN of the upstream datajob
+            downstream: URN of the downstream dataset
+        """
+        ...
+
+    @overload
+    def add_lineage(
+        self,
+        *,
+        upstream: DatajobUrnOrStr,
+        downstream: DatajobUrnOrStr,
+    ) -> None:
+        """
+        Add datajob to datajob lineage (job dependency).
+
+        Args:
+            upstream: URN of the upstream datajob
+            downstream: URN of the downstream datajob
+        """
+        ...
+
+    # The actual implementation that handles all overloaded cases
+    def add_lineage(
+        self,
+        *,
+        upstream: Union[DatasetUrnOrStr, DatajobUrnOrStr],
+        downstream: Union[DatasetUrnOrStr, DatajobUrnOrStr],
+        column_lineage: Union[
+            None, ColumnLineageMapping, Literal["auto_fuzzy", "auto_strict"]
+        ] = None,
+        query_text: Optional[str] = None,
+    ) -> None:
+        """
+        Add lineage between two entities.
+
+        This flexible method handles different combinations of entity types:
+        - Dataset to Dataset lineage (with optional column mapping and query text)
+        - Dataset to DataJob lineage (dataset as input to job)
+        - DataJob to Dataset lineage (dataset as output from job)
+        - DataJob to DataJob lineage (job dependency)
+
+        Args:
+            upstream: URN of the upstream entity (dataset or datajob)
+            downstream: URN of the downstream entity (dataset or datajob)
+            column_lineage: Optional column-level lineage mapping or auto-generation method
+                        (only applicable for dataset-to-dataset lineage)
+            query_text: Optional SQL query text that defines the transformation
+                    (only applicable for dataset-to-dataset lineage)
+
+        Raises:
+            InvalidUrnError: If the URNs provided are invalid
+            SdkUsageError: If certain parameter combinations are not supported
+        """
+        # Validate parameter combinations
+        upstream_entity_type = Urn.from_string(upstream).entity_type
+        downstream_entity_type = Urn.from_string(downstream).entity_type
+
+        # Validate parameter combinations
+        if (
+            upstream_entity_type == "dataJob" or downstream_entity_type == "dataJob"
```

Comment:
can we use `isinstance(urn, DataJobUrn)` checks instead?

Reply:
i figured out this is somehow trickly, will investigate more